codegen/src/test/java/software/amazon/awssdk/codegen/AddShapesTest.java

```diff
@@ -0,0 +1,199 @@
+import software.amazon.awssdk.codegen.model.service.Location;
+import software.amazon.awssdk.codegen.model.service.ServiceModel;
+
+class AddShapesTest {
+
+    @Test
+    void generateShapeModel_memberRequiredByShape_setsMemberModelAsRequired() {
+        String requestShapeName = "AbortMultipartUploadRequest";
+        String queryParamName = "UploadId";
+
+        ServiceModel serviceModel = serviceModel(true);
+        C2jModels c2jModels = C2jModels.builder()
+                                       .serviceModel(serviceModel)
+                                       .customizationConfig(CustomizationConfig.create())
+                                       .build();
+        IntermediateModel testModel = new IntermediateModelBuilder(c2jModels).build();
+
+        ShapeModel requestShapeModel = testModel.getShapes().get(requestShapeName);
+        MemberModel uploadIdMemberModel = requestShapeModel.findMemberModelByC2jName(queryParamName);
+
+        assertThat(requestShapeModel.getRequired()).contains(queryParamName);
+        assertThat(uploadIdMemberModel.getHttp().getLocation()).isEqualTo(Location.QUERY_STRING);
+        assertThat(uploadIdMemberModel.isRequired()).isTrue();
+    }
+
+    @Test
+    void generateShapeModel_memberNotRequiredByShape_doesNotSetMemberModelAsRequired() {
+        String requestShapeName = "AbortMultipartUploadRequest";
+        String queryParamName = "UploadId";
+
+        ServiceModel serviceModel = serviceModel(false);
+        C2jModels c2jModels = C2jModels.builder()
+                                       .serviceModel(serviceModel)
+                                       .customizationConfig(CustomizationConfig.create())
+                                       .build();
+        IntermediateModel testModel = new IntermediateModelBuilder(c2jModels).build();
+
+        ShapeModel requestShapeModel = testModel.getShapes().get(requestShapeName);
+        MemberModel uploadIdMemberModel = requestShapeModel.findMemberModelByC2jName(queryParamName);
+
+        assertThat(requestShapeModel.getRequired()).doesNotContain(queryParamName);
+        assertThat(uploadIdMemberModel.getHttp().getLocation()).isEqualTo(Location.QUERY_STRING);
+        assertThat(uploadIdMemberModel.isRequired()).isFalse();
+    }
+
+    private static ServiceModel serviceModel(boolean requireUploadId) {
+        JSON mapper = jsonMapper();
+        try {
+            return mapper.beanFrom(ServiceModel.class, serviceModelDefinition(requireUploadId));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static JSON jsonMapper() {
+        return JSON.builder()
+                   .disable(JSON.Feature.FAIL_ON_UNKNOWN_BEAN_PROPERTY)
+                   .enable(JSON.Feature.READ_JSON_ARRAYS_AS_JAVA_ARRAYS)
+                   .treeCodec(new JacksonJrsTreeCodec())
+                   .build();
+    }
+
+    private static String serviceModelDefinition(boolean requireUploadId) {
```

Comment:
The common way of providing a service model for testing is to use an actual service-2.json, most of them are stored in `resources/poet/client/c2j`. You should be able to add the trait in the query/xml folders in an operation that you define, but is configured like the method we want to apply this to for real. As you can see from the service models in this repo operations are not exact replicas of existing service models because we're both building generic features and over time as services change their models examples would deviate anyway.

Reply:
Codegen tests have been updated with new commit.